### DISCUSSION TITLE
Commands missing from command registry when workspace opened using workspace file (.theia-workspace file)

### DISCUSSION THREAD

**Discussion by svkr2k** [eclipse-theia/theia#8470] [General]
Hi all,
I'm tried opening a workspace with a large set of files in 2 ways:

* Test #1. WorkspaceService.open(folderURI)
folderURI = uri for a folder that contains source code.
This method always works fine.

* Test #2. WorkspaceService.open(workspace-file-uri)
In this approach, a workspace file is created (e.g., a file myWorkspace.theia-workspace that has paths to folders that should be opened). when using this approach, some of the commands from CommandRegistry seem to be missing, eventhough they were added. The following error is shown:
```
command.ts:268 Uncaught (in promise) Error: The command 'myTestCommand' cannot be executed. There are no active handlers available for the command.
    at CommandRegistry.<anonymous> (command.ts:268)
    at step (command.ts:15)
    at Object.next (command.ts:15)
    at command.ts:15
    at new Promise (<anonymous>)
    at ../node_modules/@theia/core/lib/common/command.js.__awaiter (command.ts:15)
    at CommandRegistry.../node_modules/@theia/core/lib/common/command.js.CommandRegistry.executeCommand (command.ts:257)
```

Has anyone seen this issue before?

**Comment by kittaakos** [comment]
Duplicating the issue won't help: https://community.theia-ide.org/t/commands-missing-from-command-registry-when-workspace-opened-using-workspace-file-theia-workspace-file/

Instead of filing the same question twice, please share a public repository with one or two examples clearly depicting the problem. Thank you!

**Reply by svkr2k** [reply]
@kittaakos , thank you for your suggestion.
I apologize, currently, I dont have a public repository to demonstrate the behavior.

But, I found a similar scenario in theia.

1. Open theia editor
2. Select File->Open Workspace... menu item under File menu
3. Choose a `myworkspace.theia-workspace` file and click OK => workspace gets opened.
4. Please note that in file navigator a root node is shown and it could have multiple folders as children
5. Select the root/workspace node ==> Pls note that the menu items `New File` and `New Folder` are disabled.
6. Now, choose one of the child folder nodes of root. ==> now, the menu items `New File` and `New Folder` are enabled.

Similarly, in my setup, some of my custom commands can't be seen when I choose the workspace node.
So, I would like to either:
1. Always make sure that root node cant be selected OR
2. Enable my custom commands that are hidden when I choose workspace/root node

Hope this helps.
Thank you, again for help.

**Reply by kittaakos** [reply]
I am sure this is a known bug 👆but could not find the corresponding issue. @vince-fugnitto, can you recall which issue was it? Maybe we have already closed it as we thought this was fixed.

**Reply by vince-fugnitto** [reply]
@svkr2k I could not reproduce your bug regarding `new file` and `new folder` in a multiple root workspace, which version of the framework are you using?

Here, we are selecting the root (workspace name) node, and performing a new file. Since no child root is selected in the mulitple-root workspace, it fallsback to creating the file/folder at the first child.

![aaa](https://user-images.githubusercontent.com/40359487/92238739-cfa64f00-ee87-11ea-867b-4371c978c5e8.gif)

> Similarly, in my setup, some of my custom commands can't be seen when I choose the workspace node.
So, I would like to either:

This is really based on your implementation, not the framework itself.

**Reply by svkr2k** [reply]
Thank you, @vince-fugnitto . I'm sorry ... using a very old version - 0.8.0 (I shall migrate to the latest framework sometime next month).

If I can know how the commands get hidden/inactive when I choose the workspace node, I can try to fix my implementation.

Also, I tried to always select the 1st child folder of workspace node/root when the ide opens => This successfully selects the 1st child node, but i still see the same issue of commands missing. If I click the 1st child node with the mouse pointer, then all my commands becomes active.

**Reply by vince-fugnitto** [reply]
> If I can know how the commands get hidden/inactive when I choose the workspace node, I can try to fix my implementation.

It depends on the implementation, I cannot tell without actually seeing your code.

For our implementation, we have `URI`-based command handlers:

https://github.com/eclipse-theia/theia/blob/636b181669bbf57cbdec0a690c7b08fd72ff1ed7/packages/workspace/src/browser/workspace-commands.ts#L345-L355

For example, the `WorkspaceRootUriAwareCommandHandler` uses the following to determine the `uri` for which to perform the actions:

https://github.com/eclipse-theia/theia/blob/636b181669bbf57cbdec0a690c7b08fd72ff1ed7/packages/workspace/src/browser/workspace-commands.ts#L514-L524

This is the behavior I expressed and demonstrated, the first workspace child root is used when no selection is present.
